### PR TITLE
Lims 3710: Orders: Archived Test unit: Archive Approach: Archived test units shouldn't appear in orders in the drop down list

### PR DESCRIPTION
**Pull request by dianaeldemerdash**
lims/3710

### CODE REVIEW

**Comment by dianaeldemerdash** [comment]
@0xIslamTaha second report is green and no conflict so the problem now I think in the build but I didn't find restart the build in the report to start it

File: ui_testing/testcases/basic_tests/test06_orders.py [hunk 1]
```diff
@@ -2118,3 +2099,39 @@ def test033_Duplicate_main_order_and_cahange_materiel_type(self):
...
+        self.base_selenium.LOGGER.info('select random test unit record')
+        testunits, payload = self.test_unit_api.get_all_test_units()
+        random_row = random.choice(testunits['testUnits'])
+
+        testunit_name = random_row['name']
+        self.order_page.apply_filter_scenario(filter_element='test_units:testunit_name_filter', filter_text=testunit_name,
+                                              field_type='text')
+        row = self.test_unit_page.get_last_test_unit_row()
+        self.order_page.click_check_box(source=row)
+        self.test_unit_page.archive_selected_test_units()
```
Comment (ghada-37) [inline ui_testing/testcases/basic_tests/test06_orders.py:2119]: can't we replace all above steps by define new function in test_unit_api that list all archived test units?

File: ui_testing/pages/order_page.py [hunk 2]
```diff
@@ -521,3 +521,28 @@ def remove_article(self, testplans=''):
             self.base_selenium.click(element='general:confirmation_button')
         self.sleep_small()
 
+    def set_random_test_unit(self, test_units='', sub_order_index=''):
```
Comment (ghada-37) [inline ui_testing/pages/order_page.py:524]: why do you need to define new function?
you can use https://github.com/modeso-open-source/1lims-automation/blob/b278dfc150ae4a54a0fdb25d1ea39f33c1081341/ui_testing/pages/order_page.py#L372

File: ui_testing/testcases/basic_tests/test06_orders.py (hunk 1 above)
Comment (dianaeldemerdash) [reply ui_testing/testcases/basic_tests/test06_orders.py:2119]: done

File: ui_testing/pages/order_page.py (hunk 2 above)
Comment (dianaeldemerdash) [reply ui_testing/pages/order_page.py:524]: @ghada-37 because this method will not work because I need to use set_test _unit  to set specific index and the set_test_unit in this method will not do that  ( for testunit in test_units:
            self.set_test_unit(test_units=testunit) )

File: ui_testing/testcases/basic_tests/test06_orders.py [hunk 3]
```diff
@@ -2118,3 +2099,30 @@ def test033_Duplicate_main_order_and_cahange_materiel_type(self):
...
+        Orders: Archived Test unit: Archive Approach: Archived test units shouldn't appear in orders in the drop down list
+        LIMS-3710
+        :return:
+        """
+        testunits, payload = self.test_unit_api.get_all_test_units(limit=20, deleted=1)
+        random_row = random.choice(testunits['testUnits'])
+
+        self.order_page.get_orders_page()
+        self.base_selenium.LOGGER.info('select random order record')
+        orders, payload = self.orders_api.get_all_orders(limit=20)
```
Comment (ghada-37) [inline ui_testing/testcases/basic_tests/test06_orders.py:2114]: use self.info() directly

File: ui_testing/testcases/basic_tests/test06_orders.py [hunk 4]
```diff
@@ -2118,3 +2099,30 @@ def test033_Duplicate_main_order_and_cahange_materiel_type(self):
...
+        order_row = random.choice(orders['orders'])
+
+        order_no = order_row['orderNo']
+        self.order_page.apply_filter_scenario(filter_element='orders:filter_order_no', filter_text=order_no,
+                                              field_type='text')
+
+        row = self.order_page.get_last_order_row()
+        self.orders_page.open_edit_page(row)
```
Comment (ghada-37) [inline ui_testing/testcases/basic_tests/test06_orders.py:2122]: u can use this directly to save time
`self.orders_page.get_order_edit_page_by_id(id=order['id'])`

File: ui_testing/pages/order_page.py [hunk 5]
```diff
@@ -521,3 +521,28 @@ def remove_article(self, testplans=''):
...
+        suborder_elements_dict = self.base_selenium.get_row_cells_id_dict_related_to_header(
+            row=suborder_row, table_element='order:suborder_table')
+        suborder_row.click()
+        if test_units:
+            self.base_selenium.select_item_from_drop_down(
+                element='order:test_unit', item_text=test_units)
+        else:
+            self.base_selenium.select_item_from_drop_down(
+                element='order:test_unit')
+            return self.get_random_test_unit()
```
Comment (ghada-37) [inline ui_testing/pages/order_page.py:537]: it should be self.get_test_unit()

File: ui_testing/pages/order_page.py [hunk 6]
```diff
@@ -521,3 +521,28 @@ def remove_article(self, testplans=''):
...
+
+    def get_random_test_unit(self, sub_order_index =''):
```
Comment (ghada-37) [inline ui_testing/pages/order_page.py:539]: I found no need to define this new function, unless u need to get test unit of specific suborder

File: ui_testing/testcases/basic_tests/test06_orders.py [hunk 7]
```diff
@@ -26,6 +28,8 @@ def setUp(self):
         self.article_api = ArticleAPI()
         self.test_unit_api = TestUnitAPI()
         self.contacts_api = ContactsAPI()
+        self.test_unit_page = TstUnits()
+        self.test_units_page = TstUnit()
```
Comment (ghada-37) [inline ui_testing/testcases/basic_tests/test06_orders.py:32]: you don't used them, remove them
and if you need to them define them in test case itself for setup time save

File: ui_testing/testcases/basic_tests/test06_orders.py [hunk 8]
```diff
@@ -2118,3 +2099,30 @@ def test033_Duplicate_main_order_and_cahange_materiel_type(self):
...
+        order_row = random.choice(orders['orders'])
+
+        order_no = order_row['orderNo']
+        self.order_page.apply_filter_scenario(filter_element='orders:filter_order_no', filter_text=order_no,
+                                              field_type='text')
+
+        row = self.order_page.get_last_order_row()
+        self.orders_page.open_edit_page(row)
+
+        result = self.order_page.set_random_test_unit(test_units=random_row['name'], sub_order_index=0)
```
Comment (ghada-37) [inline ui_testing/testcases/basic_tests/test06_orders.py:2124]: self.order_page.set_random_test_unit(test_units=random_row['name'], sub_order_index=0) this function don't return anything in case test_units has a value !

File: ui_testing/pages/order_page.py (hunk 2 above)
Comment (ghada-37) [reply ui_testing/pages/order_page.py:524]: I still found u can use it 
```
update_suborder(sub_order_index=0,test_units=[random_row['name']])

File: ui_testing/testcases/basic_tests/test06_orders.py (hunk 3 above)
Comment (dianaeldemerdash) [reply ui_testing/testcases/basic_tests/test06_orders.py:2114]: done

File: ui_testing/testcases/basic_tests/test06_orders.py (hunk 4 above)
Comment (dianaeldemerdash) [reply ui_testing/testcases/basic_tests/test06_orders.py:2122]: How can I save time by this one

File: ui_testing/testcases/basic_tests/test06_orders.py (hunk 7 above)
Comment (dianaeldemerdash) [reply ui_testing/testcases/basic_tests/test06_orders.py:32]: done

File: ui_testing/testcases/basic_tests/test06_orders.py (hunk 4 above)
Comment (ghada-37) [reply ui_testing/testcases/basic_tests/test06_orders.py:2122]: It opens url of edit page directly without geting orders page and waiting its loading then filter and wait result

File: ui_testing/testcases/basic_tests/test06_orders.py (hunk 8 above)
Comment (dianaeldemerdash) [reply ui_testing/testcases/basic_tests/test06_orders.py:2124]: Done

File: ui_testing/pages/order_page.py (hunk 6 above)
Comment (dianaeldemerdash) [reply ui_testing/pages/order_page.py:539]: done

File: ui_testing/pages/order_page.py (hunk 5 above)
Comment (dianaeldemerdash) [reply ui_testing/pages/order_page.py:537]: done

File: ui_testing/testcases/basic_tests/test06_orders.py (hunk 4 above)
Comment (dianaeldemerdash) [reply ui_testing/testcases/basic_tests/test06_orders.py]: @ghada-37 I got your point now

File: ui_testing/testcases/basic_tests/test06_orders.py [hunk 9]
```diff
@@ -2128,7 +2104,27 @@ def test033_Duplicate_main_order_and_cahange_materiel_type(self):
...
+    def test035_archived_test_unit_shoudnt_display_in_the_order_drop_down_list(self):
+        """
+        Orders: Archived Test unit: Archive Approach: Archived test units shouldn't appear in orders in the drop down list
+        LIMS-3710
+        :return:
+        """
+        testunits, payload = self.test_unit_api.get_all_test_units(limit=20, deleted=1)
+        random_row = random.choice(testunits['testUnits'])
+
+        self.order_page.get_orders_page()
```
Comment (ghada-37) [inline ui_testing/testcases/basic_tests/test06_orders.py:2117]: Remove this line

File: ui_testing/testcases/basic_tests/test06_orders.py [hunk 10]
```diff
@@ -2128,7 +2104,27 @@ def test033_Duplicate_main_order_and_cahange_materiel_type(self):
...
+        self.info('select random order record')
+        orders, payload = self.orders_api.get_all_orders(limit=20)
+        order_row = random.choice(orders['orders'])
+
+        self.orders_page.get_order_edit_page_by_id(id=order_row['id'])
+
+        result = self.order_page.update_suborder(test_units=[random_row['name']], sub_order_index=0)
```
Comment (ghada-37) [inline ui_testing/testcases/basic_tests/test06_orders.py:2124]: Update suborder has no "return" statement in it 
So "result" value  not defined
Also, you need to make assertion that after set test unit , get test unit returns null

Comment (dianaeldemerdash) [reply ui_testing/testcases/basic_tests/test06_orders.py:2124]: @0xIslamTaha plesae Islam I wan your help here we have same as this test case and already merged on the master branch so I want here to search on archived test units and it should't appear in the drop down list same as this test case in the roles and permissions section same as here  
        result = self.header_page.set_user_role(user_role=role_random_name)
        self.assertFalse(result, 'no results found ') so this is correct

File: ui_testing/testcases/basic_tests/test06_orders.py (hunk 9 above)
Comment (0xIslamTaha) [reply ui_testing/testcases/basic_tests/test06_orders.py:2117]: @dianaeldemerdash

File: ui_testing/testcases/basic_tests/test06_orders.py [hunk 11]
```diff
@@ -2128,7 +2104,27 @@ def test033_Duplicate_main_order_and_cahange_materiel_type(self):
...
-        
+
+    def test035_archived_test_unit_shoudnt_display_in_the_order_drop_down_list(self):
+        """
+        Orders: Archived Test unit: Archive Approach: Archived test units shouldn't appear in orders in the drop down list
+        LIMS-3710
+        :return:
+        """
+        testunits, payload = self.test_unit_api.get_all_test_units(limit=20, deleted=1)
+        random_row = random.choice(testunits['testUnits'])
```
Comment (0xIslamTaha) [inline ui_testing/testcases/basic_tests/test06_orders.py:2115]:
```suggestion
        test_unit = random.choice(testunits['testUnits'])
```

File: ui_testing/testcases/basic_tests/test06_orders.py (hunk 10 above)
Comment (0xIslamTaha) [reply ui_testing/testcases/basic_tests/test06_orders.py:2124]: @dianaeldemerdash 
- If I merged the same thing before, then It is my fault and please send it to me and I'm gonna fix it.

- I debugged it and I can see that @ghada-37 already mentioned the correct behavior and here is the details:
As u can see from this [implementation](https://github.com/modeso-open-source/1lims-automation/blob/0a181e3961b74159c0b49ce8dc9d71882b94377e/ui_testing/pages/order_page.py#L399), there is no return value in your call and if u tracked the code u can see that `result` has no value so `self.assertFalse(result)` will always pass.

image

- As a suggestion, here u need to create a new function to `check_if_test_unit_existing` or not and it should return false in this case

File: ui_testing/pages/order_page.py (hunk 2 above)
Comment (dianaeldemerdash) [reply ui_testing/pages/order_page.py]: Done

File: ui_testing/testcases/basic_tests/test06_orders.py (hunk 10 above)
Comment (dianaeldemerdash) [reply ui_testing/testcases/basic_tests/test06_orders.py]: Done

File: ui_testing/testcases/basic_tests/test06_orders.py (hunk 11 above)
Comment (dianaeldemerdash) [reply ui_testing/testcases/basic_tests/test06_orders.py]: Done

File: ui_testing/testcases/basic_tests/test06_orders.py (hunk 9 above)
Comment (dianaeldemerdash) [reply ui_testing/testcases/basic_tests/test06_orders.py:2121]: Done

**Comment by 0xIslamTaha** [comment]
@dianaeldemerdash 
This test case could raise a false failure when it gets a random archived test unit with the name "Xyz" and in the same time, there is another active test unit with the same name so `self.order_page.is_testunit_existing(
            test_unit="6e3811219b")` will return true

as an example: there are two test units with `6e3811219b` name one active and the other is archived

**Comment by dianaeldemerdash** [comment]
@0xIslamTaha yes it is a valid point because the test unit name is not a unique so I will create test unit using the api then archive it I think it will be a solution or do you recommend another solution

**Comment by 0xIslamTaha** [comment]
> @0xIslamTaha yes it is a valid point because the test unit name is not a unique so I will create test unit using the api then archive it I think it will be a solution or do you recommend another solution

@dianaeldemerdash This should do it but make sure u are using API to create and archive it

**Comment by dianaeldemerdash** [comment]
@0xIslamTaha I have a question here regarding this point now I did the following 
1-create test unit using api ( payload = self.test_unit_api.create_qualitative_testunit()[0]['testUnit'] ) 
& this will return the testUnit with this form {'No': 28054, 'testUnitId': 1323}
2-then archive this record using api ( self.test_unit_api.archive_testunits(ids=payload['testUnitId']))
to give the archive funtion the test unit api 
but the problem now I thing we have a wrong way that when we give the archive method the id like this way payload['testUnitId'] it gave me this error (can only join an iterable)
I tried a lot of times to find the problem but I don't know so what is your your solution in that @

**Comment by 0xIslamTaha** [comment]
> @0xIslamTaha I have a question here regarding this point now I did the following
> 1-create test unit using api ( payload = self.test_unit_api.create_qualitative_testunit()[0]['testUnit'] )
> & this will return the testUnit with this form {'No': 28054, 'testUnitId': 1323}
> 2-then archive this record using api ( self.test_unit_api.archive_testunits(ids=payload['testUnitId']))
> to give the archive funtion the test unit api
> but the problem now I thing we have a wrong way that when we give the archive method the id like this way payload['testUnitId'] it gave me this error (can only join an iterable)
> I tried a lot of times to find the problem but I don't know so what is your your solution in that @

You have to do something like this 

```python
re, pay =  self.test_unit_api.create_qualitative_testunit()
response, payload = self.test_unit_api.archive_testunits(ids=[str(re['testUnit']['testUnitId'])])
```

**Comment by dianaeldemerdash** [comment]
nosetests3 -vs --logging-level=WARNING ui_testing/testcases/basic_tests/test06_orders.py:OrdersTestCases.test035_archived_test_unit_shoudnt_display_in_the_order_drop_down_list --tc-file=config.ini --tc=site.password:admin 
2020-06-08 20:14:54.046 | INFO     | api_testing.apis.base_api:info:76 - Get authorized api session.
2020-06-08 20:14:54.739 | INFO     | api_testing.apis.base_api:info:76 - session ID : eyJhbGciOi .....
Orders: Archived Test unit: Archive Approach: Archived test units shouldn't appear in orders in the drop down list ...  
2020-06-08 20:14:54.960 | INFO     | ui_testing.testcases.base_test:info:135 - Test case : test035_archived_test_unit_shoudnt_display_in_the_order_drop_down_list
2020-06-08 20:15:36.079 | INFO     | ui_testing.pages.base_pages:wait_until_page_is_loaded:537 - wait until page is loaded
2020-06-08 20:15:36.144 | INFO     | api_testing.apis.base_api:info:76 - GET : https://automation.1lims.com/api/testUnits
2020-06-08 20:15:37.429 | INFO     | api_testing.apis.base_api:info:76 - Status code: 1
2020-06-08 20:15:37.430 | INFO     | api_testing.apis.base_api:info:76 - GET : https://automation.1lims.com/api/testUnits/1335/archive
2020-06-08 20:15:37.964 | INFO     | api_testing.apis.base_api:info:76 - Status code: 1
2020-06-08 20:15:41.359 | INFO     | ui_testing.pages.base_pages:wait_until_page_is_loaded:537 - wait until page is loaded
2020-06-08 20:15:42.522 | INFO     | ui_testing.pages.order_page:set_new_order:18 - Set new order.
2020-06-08 20:15:50.102 | INFO     | ui_testing.pages.base_pages:sleep_small:43 -  Small sleep.
2020-06-08 20:16:05.795 | INFO     | ui_testing.testcases.base_test:info:135 - Asset test unit is not existing in the list
2020-06-08 20:16:15.231 | INFO     | ui_testing.pages.base_selenium:select_item_from_drop_down:454 -  There is no afd2fa71ee option in the drop-down
2020-06-08 20:16:15.508 | INFO     | ui_testing.testcases.base_test:info:135 - TearDown.        
ok

----------------------------------------------------------------------
Ran 1 test in 80.549s

OK

File: ui_testing/testcases/basic_tests/test06_orders.py (hunk 9 above)
Comment (0xIslamTaha) [reply ui_testing/testcases/basic_tests/test06_orders.py:2043]: @dianaeldemerdash seems u didn't remove it yet

File: ui_testing/testcases/basic_tests/test06_orders.py [hunk 12]
```diff
@@ -2031,6 +2032,24 @@ def test033_Duplicate_main_order_and_cahange_materiel_type(self):
...
         self.assertEqual(suborder_data['Test Plans'], selected_test_plan['testPlanName'])
 
+    def test035_archived_test_unit_shoudnt_display_in_the_order_drop_down_list(self):
+        """
+        Orders: Archived Test unit: Archive Approach: Archived test units shouldn't appear in orders in the drop down list
+        LIMS-3710
+        :return:
+        """
+        re, payload = self.test_unit_api.create_qualitative_testunit()
+        archived_testunit = self.test_unit_api.archive_testunits(ids=[str(re['testUnit']['testUnitId'])])
```
Comment (0xIslamTaha) [inline ui_testing/testcases/basic_tests/test06_orders.py:2042]:
```suggestion
       self.test_unit_api.archive_testunits(ids=[str(re['testUnit']['testUnitId'])])
```

File: ui_testing/testcases/basic_tests/test06_orders.py [hunk 13]
```diff
@@ -2031,6 +2032,24 @@ def test033_Duplicate_main_order_and_cahange_materiel_type(self):
...
+        self.order_page.get_orders_page()
+        self.base_selenium.click(element='orders:new_order')
+        self.order_page.set_new_order()
+        self.order_page.sleep_small()
+        self.order_page.set_material_type_of_first_suborder(material_type='r', sub_order_index=0)
```
Comment (0xIslamTaha) [inline ui_testing/testcases/basic_tests/test06_orders.py:2047]: You should use the material type that has been used in creating the test unit. Right?

Comment (dianaeldemerdash) [reply ui_testing/testcases/basic_tests/test06_orders.py:2047]: @0xIslamTaha No,  the test unit creation using the api, it set the material type by default to (All ) but In the test plans and the orders we don't have all material type so this mean If I select any material type for sure this test unit appear because I created it with ( All)

File: ui_testing/testcases/basic_tests/test06_orders.py (hunk 12 above)
Comment (dianaeldemerdash) [reply ui_testing/testcases/basic_tests/test06_orders.py:2042]: Done

**Comment by dianaeldemerdash** [comment]
nosetests3 -vs --logging-level=WARNING ui_testing/testcases/basic_tests/test06_orders.py:OrdersTestCases.test035_archived_test_unit_shoudnt_display_in_the_order_drop_down_list --tc-file=config.ini --tc=site.password:admin 
2020-06-10 03:42:19.568 | INFO     | api_testing.apis.base_api:info:76 - Get authorized api session.
2020-06-10 03:42:20.356 | INFO     | api_testing.apis.base_api:info:76 - session ID : eyJhbGciOi .....
Orders: Archived Test unit: Archive Approach: Archived test units shouldn't appear in orders in the drop down list ...  
2020-06-10 03:42:20.580 | INFO     | ui_testing.testcases.base_test:info:135 - Test case : test035_archived_test_unit_shoudnt_display_in_the_order_drop_down_list
2020-06-10 03:42:52.018 | INFO     | ui_testing.pages.base_pages:wait_until_page_is_loaded:537 - wait until page is loaded
2020-06-10 03:42:53.403 | INFO     | api_testing.apis.base_api:info:76 - GET : https://automation.1lims.com/api/testUnits
2020-06-10 03:42:54.254 | INFO     | api_testing.apis.base_api:info:76 - Status code: 1
2020-06-10 03:42:54.255 | INFO     | api_testing.apis.base_api:info:76 - GET : https://automation.1lims.com/api/testUnits/1361/archive
2020-06-10 03:42:54.804 | INFO     | api_testing.apis.base_api:info:76 - Status code: 1
2020-06-10 03:42:55.377 | INFO     | ui_testing.pages.order_page:set_new_order:18 - Set new order.
2020-06-10 03:43:03.037 | INFO     | ui_testing.pages.base_pages:sleep_small:43 -  Small sleep.
2020-06-10 03:43:18.902 | INFO     | ui_testing.testcases.base_test:info:135 - Asset test unit is not existing in the list
2020-06-10 03:43:28.480 | INFO     | ui_testing.pages.base_selenium:select_item_from_drop_down:454 -  There is no 406de5c1f4 option in the drop-down
2020-06-10 03:43:28.894 | INFO     | ui_testing.testcases.base_test:info:135 - TearDown.        
ok

----------------------------------------------------------------------
Ran 1 test in 68.315s

OK
